pkg/alerter/alerter.go

```diff
@@ -0,0 +1,245 @@
+	"github.com/GoogleCloudPlatform/testgrid/pkg/summarizer"
+	"github.com/GoogleCloudPlatform/testgrid/util/gcs"
+	"github.com/golang/protobuf/proto"
+	"github.com/sirupsen/logrus"
+)
+
+// Read summary protos and call functions for each Klaxon on certain conditions
+// Maintains a copy of summary protos to recognize last read state after reboot
+func Update(ctx context.Context, client *storage.Client, path gcs.Path, concurrency int, dashboard string, subscribers []Klaxon, confirm bool) error {
+	if concurrency < 1 {
+		return fmt.Errorf("concurrency must be positive, got: %d", concurrency)
+	}
+
+	if len(subscribers) == 0 {
+		logrus.Warn("No subscribers to alerter; events will drop silently")
+	}
+
+	// get config
+	cfg, err := config.ReadGCS(ctx, client.Bucket(path.Bucket()).Object(path.Object()))
+	if err != nil {
+		return fmt.Errorf("Failed to read config: %w", err)
+	}
+	logrus.Infof("Found %d dashboards", len(cfg.Dashboards))
+
+	dashboards := make(chan *configpb.Dashboard)
+	var wg sync.WaitGroup
+
+	errCh := make(chan error)
+
+	// Launch monitoring goroutines
+	for i := 0; i < concurrency; i++ {
+		wg.Add(1)
+		go func() {
+			for dash := range dashboards {
+				// Read the new summary file from the expected location (summary)
+				// If it's not there, alerts can't be generated for this dashboard
+				newPath, err := path.ResolveReference(&url.URL{Path: summaryPath(dash.Name)})
+				if err != nil {
+					logrus.Errorf("Could not resolve path to current summaries of %s", dash.Name)
+					errCh <- err
+					continue
+				}
+
+				newSummary, err := ReadSummary(ctx, client, *newPath)
+				if err != nil {
+					logrus.Errorf("Could not read summary from %v", newPath)
+					errCh <- err
+					continue
+				}
+
+				// Read the old summary file from the expected location (alerter)
+				// It may not be there the first time this is run
+				oldPath, err := path.ResolveReference(&url.URL{Path: alerterPath(dash.Name)})
+				if err != nil {
+					logrus.Errorf("Could not resolve path to previous summaries of %s", dash.Name)
+					errCh <- err
+					continue
+				}
+
+				oldSummary, err := ReadSummary(ctx, client, *oldPath)
+				if err != nil {
+					logrus.Warnf("Could not read summary from %v; trying write", oldPath)
+				}
+
+				if !confirm {
+					logrus.Infof("Summarizers read for dashboard %s (dry-run)", dash.Name)
+					continue
+				}
+
+				// Detect changes and call the Klaxon function(s) if those changes are pertinent
+				if oldSummary != nil {
+					callOutDifferences(ctx, newSummary, oldSummary, dash, subscribers)
+				}
+
+				// Write the new summary to where the old summary was
+				err = summarizer.WriteSummary(ctx, client, *oldPath, newSummary)
+				if err != nil {
+					logrus.Errorf("Could not write summary to %v", oldPath)
+					errCh <- err
+					continue
+				}
+				errCh <- nil
+			}
+			wg.Done()
+		}()
+	}
+
+	// Subroutine to concatenate errors together
+	resultCh := make(chan error)
+	go func() {
```

Comment:
Same thing with closing the channel (use a defer as soon as you enter this function, telling me I don't have to scan this method for cases where you accidentally return early without closing)